cts.json

```diff
@@ -3819,6 +3819,48 @@
       "selector": "$[-1:-10:-231584178474632390847141970017375815706539969331281128078915168015826259279872]",
       "invalid_selector": true
     },
+    {
+      "name": "functions, arguments, parenthesized expression",
+      "selector": "$[?match((@.a), ('a.*'))]",
```

Comment:
I'm not sure this is correct. Here's how I see it. The parens around the arguments can only be used in a `logical-exp`, which building blocks (apart from the parens, `&&`, `||`) are `comparison-expr` and `test-expr`. So `@.a` and `a.*` are `test-expr` meaning they are `LogicalType`. But `match()` only accepts `ValueType` arguments, therefore the expression is not well-typed and the implementation must raise an error.

Reply:
I see where you're going with this. I'm mobile right now, but it can be tested at https://json-everything.net/json-path.

Reply:
@gregsdennis 
![image](https://github.com/jsonpath-standard/jsonpath-compliance-test-suite/assets/831399/90b672fc-2c8e-4851-b4bc-d1843d73862c)

Reply:
One can argue that it could be better not to allow `logical-expr` in
```abnf
function-argument   = literal /
                      filter-query / ; (includes singular-query)
                      logical-expr /
                      function-expr
 ```

Reply:
FWIW my implementation did not allow parens in function args until now. This is a useful test, there's just no such standard function that can be used in it. In my implementation I introduced a non-standard logical function to have this case covered.

Reply:
> One can argue that it could be better not to allow logical-expr in...

We don't want to preclude the possibility of someone registering such a function if it proves useful.

Reply:
> `$[?match((@.a), ('a.*'))]`

```
$ pbpaste | jpt -
/Users/cabo/lib/ruby/gems/3.2.0/gems/jpt-0.1.2/lib/jpt.rb:84:in `from_jp': Expected one of " ", "\t", "\n", "\r", "==", "!=", "<=", ">=", "<", ">" at line 1, column 23 (byte 23) after [?match((@.a), ('a.*' (RuntimeError)
Expected one of " ", "\t", "\n", "\r", "==", "!=", "<=", ">=", "<", ">" at line 1, column 23 (byte 23):
$[?match((@.a), ('a.*'))]
~~~~~~~~~~~~~~~~~~~~~~^
```
Clearly, ('a.*') is not a logical-expr, so the ABNF grammar forbids this.

> `$[?match((@.a), 'a.*')]`

```
$ pbpaste | jpt -
["$", ["filt", ["func", "match", ["@", "a"], "a.*"]]]
```

Reply:
My implementation still doesn't like `$[?match((@.a), 'a.*')]`.  Removing the parentheses around `@.a` works, though.

I think @f3ath's analysis is correct that `(@.a)` is interpreted as a general expression (yielding `LogicalType`).  Specifically, it doesn't meet the requirements of a singular query.

The relevant spec section is [2.4.3](https://www.ietf.org/archive/id/draft-ietf-jsonpath-base-21.html#section-2.4.3), toward the bottom:

> (2) The arguments of the function expression are well-typed when each argument of the function can be used for the declared type of the corresponding parameter, according to one of the following conditions:
> 
> - ...
> - When the declared type of the parameter is ValueType and the argument is one of the following:
>     - A value expressed as a literal.
>     - A singular query. In this case:
>       - If the query results in a nodelist consisting of a single node, the argument is the value of the node.
>       - If the query results in an empty nodelist, the argument is the special result Nothing.

Now, it could be argued that `(@.a)` can be determined to be a singular query _at runtime_, after the query runs and results are returned, but simple syntactic analysis doesn't allow `(@.a)` where a `ValueType` is expected.

However, making this argument means that the query's well-typedness can't be determined until runtime, which (to me) seems to violate section [2.1](https://www.ietf.org/archive/id/draft-ietf-jsonpath-base-21.html#section-2.1):

> A JSONPath implementation MUST raise an error for any query which is not well-formed and valid. The well-formedness and the validity of JSONPath queries are independent of the JSON value the query is applied to. **No further errors relating to the well-formedness and the validity of a JSONPath query can be raised during application of the query to a value.** This clearly separates well-formedness/validity errors in the query from mismatches that may actually stem from flaws in the data.

Reply:
@gregsdennis Can you please clarify why the well-typedness can't be determined at compile time? Given a function whose arguments' types are known at compile type, that each argument is exactly one of `ValueType`, `NodesType`, or ` LogicalType`, and considering the type (Nodes to Value and Nodes to Logical) it should be possible IMO.

Reply:
I was just saying that if you were to wait until runtime to determine if a query were singular, then you wouldn't be able to determine well-typedness until runtime, either, since well-typedness can depend on whether a query is singular.

---

The primary problem is that `(@.a)` is _syntactically_ a filter expression.  Expressions return `LogicalType`.  (In the context of an expression, a path on its own is an existence test, which returns `LogicalType`.)

Furthermore, we don't (and probably shouldn't) have a conversion _from_ `LogicalType`.

This means that the expression isn't well-formed as `match()` isn't receiving correctly-typed arguments.

Reply:
Ah, understood.

Reply:
@cabo does this mean that the test you [showed](https://github.com/jsonpath-standard/jsonpath-compliance-test-suite/pull/51#discussion_r1335207281) has a bug, or am interpreting this wrong?

Reply:
Well, the problem is that a reasonable parser implementation will discard the () around `@.a`, so the fact that this was syntactically a logicaltype is not visible to the semantic processing.

If I force this by doing one of

``` 
$[?match(!(@.a), 'a.*')]
$[?match(!@.a, 'a.*')]
$[?match((@.a && @.a), 'a.*')]
```

I get one of

```
*** Cannot use ["not", ["@", "a"]] with declared_type logical for required type value in ["func", "match", ["not", ["@", "a"]], "a.*"]
*** Cannot use ["and", ["@", "a"], ["@", "a"]] with declared_type logical for required type value in ["func", "match", ["and", ["@", "a"], ["@", "a"]], "a.*"]
```

Giving pure parentheses a type-coercing semantics is a flaw that overly complicated languages such as C++ have, but that I didn't know we had put into JSONPath.  Now I have to think how to integrate this into jpt's parser/ast generator.

Reply:
> reasonable parser implementation will discard the () around `@.a`

I'm not convinced of this.  Reasonability is subjective.

Reply:
s/reasonable parser/parser written by someone who has done expression language parsers before/

The AST gets a bit ugly when you need to include the parens as a node.
Of course, it can be done.